### PR TITLE
Unnest array of rows based on unnestArrayOfRows

### PR DESCRIPTION
**Pull request by aditi-pandit**
Velox unnests array of row type into a single column of the
element row type. Presto also supports unnesting such arrays
into multiple columns one for each child type of element row
type. Adds unnestArrayOfRows to add such unnesting support
in Unnest PlanNode.

Needed for https://github.com/prestodb/presto/issues/20643

### CODE REVIEW

File: velox/exec/Unnest.cpp [hunk 1]
```diff
@@ -220,13 +220,52 @@ RowVectorPtr Unnest::generateOutput(
...
+        // The array row elements could be encoded however. In that case, each
+        // child array accessed needs to be encoded with the array row elements
+        // encoding and then wrapped with the un-nest encoding for final output.
+        DecodedVector rowDecoded(*elements);
+        auto rowVector = rowDecoded.base()->as<RowVector>();
+
+        auto rowIsIdentity = rowDecoded.isIdentityMapping();
+        auto numRowElements = rowDecoded.size();
+
+        // These need to be constructed because the DecodedVector indices/nulls
```
Comment (mbasmanova) [inline velox/exec/Unnest.cpp:237]: Would you check if DecodedVector::wrap API can be used here?

Comment (aditi-pandit) [reply velox/exec/Unnest.cpp]: Thanks @mbasmanova. DecodedVector.wrap works as well.

**Comment by aditi-pandit** [comment]
@mbasmanova : Updated the code. PTAL.

File: velox/core/PlanNode.h [hunk 2]
```diff
@@ -1897,29 +1897,60 @@ class LimitNode : public PlanNode {
...
+  /// columns, one for each child type of the row. However, with legacyUnnest,
+  /// array of rows (like other array types) are expanded to a single column of
+  /// the row type.
   UnnestNode(
       const PlanNodeId& id,
       std::vector<FieldAccessTypedExprPtr> replicateVariables,
       std::vector<FieldAccessTypedExprPtr> unnestVariables,
       const std::vector<std::string>& unnestNames,
       const std::optional<std::string>& ordinalityName,
+      bool legacyUnnest,
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1933]: > legacyUnnest

Unlike Presto, Velox is a relatively new, hence, it is a bit strange to see a parameter named 'legacyUnnest' here. Also, this name is not very descriptive. 

I wonder if we can come up with a better name.

File: velox/core/PlanNode.cpp [hunk 3]
```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
...
   int unnestIndex = 0;
+  auto unnestNamesLength = unnestNames.size();
   for (const auto& variable : unnestVariables_) {
-    if (variable->type()->isArray()) {
-      names.emplace_back(unnestNames[unnestIndex++]);
-      types.emplace_back(variable->type()->asArray().elementType());
-    } else if (variable->type()->isMap()) {
-      const auto& mapType = variable->type()->asMap();
-
+    auto type = variable->type();
```
Comment (mbasmanova) [inline velox/core/PlanNode.cpp:858]: const &

File: velox/core/PlanNode.cpp [hunk 4]
```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
...
+    if (type->isArray()) {
+      auto elementType = type->asArray().elementType();
```
Comment (mbasmanova) [inline velox/core/PlanNode.cpp:860]: const &

File: velox/core/PlanNode.cpp [hunk 5]
```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
...
+      if (elementType->isRow() && !legacyUnnest_) {
+        // Array of rows are flattened to a list of columns. There is one
+        // column for each child type of the row.
+        // Note : In legacy unnest however, array of rows like other arrays
+        // are flattened to a single column of the array type.
+        auto rowType = asRowType(elementType);
```
Comment (mbasmanova) [inline velox/core/PlanNode.cpp:866]: not need for asRowType; Type base class have size() and childAt() APIs, hence, you can use elementType directly.

File: velox/core/PlanNode.cpp [hunk 6]
```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
...
+        auto numChildren = rowType->size();
+        for (auto j = 0; j < numChildren; j++) {
```
Comment (mbasmanova) [inline velox/core/PlanNode.cpp:868]: j -> i

File: velox/core/PlanNode.cpp [hunk 7]
```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
...
       types.emplace_back(mapType.keyType());
 
+      indexCheck(unnestIndex, unnestNamesLength);
       names.emplace_back(unnestNames[unnestIndex++]);
       types.emplace_back(mapType.valueType());
     } else {
       VELOX_FAIL(
           "Unexpected type of unnest variable. Expected ARRAY or MAP, but got {}.",
-          variable->type()->toString());
+          type->toString());
```
Comment (mbasmanova) [inline velox/core/PlanNode.cpp:891]: Can you also change this to VELOX_USER_FAIL?

File: velox/core/PlanNode.cpp [hunk 8]
```diff
@@ -895,6 +926,8 @@ folly::dynamic UnnestNode::serialize() const {
   if (withOrdinality_) {
     obj["ordinalityName"] = outputType()->names().back();
   }
+
+  obj["legacyUnnest"] = legacyUnnest_;
```
Comment (mbasmanova) [inline velox/core/PlanNode.cpp:930]: We need to include this property in the output of UnnestNode::toString.

File: velox/docs/develop/operators.rst [hunk 9]
```diff
@@ -665,9 +667,11 @@ number starting with 1.
    * - unnestVariables
      - Input columns of type array or map to expand.
    * - unnestNames
-     - Names to use for expanded columns. One name per array column. Two names per map column.
+     - Names to use for expanded columns. One name per array column. Two names per map column. If array of rows (and not legacyUnnest), then one column per child type of the row.
    * - ordinalityName
      - Optional name for the ordinality column.
+   * - legacyUnnest
+     - Array of rows are typically expanded to multiple columns, one for each child type of the row. However, with legacyUnnest, array of rows (like other array types) are expanded to a single column of the row type.
```
Comment (mbasmanova) [inline velox/docs/develop/operators.rst:674]: perhaps, legacyUnnest -> flattenArrayOfRow or unnestArrayOfRow

File: velox/exec/Unnest.cpp [hunk 10]
```diff
@@ -182,8 +183,7 @@ RowVectorPtr Unnest::generateOutput(
     BufferPtr elementIndices = allocateIndices(numElements, pool());
     auto* rawElementIndices = elementIndices->asMutable<vector_size_t>();
 
-    auto nulls =
-        AlignedBuffer::allocate<bool>(numElements, pool(), bits::kNotNull);
+    auto nulls = allocateNulls(numElements, pool());
```
Comment (mbasmanova) [inline velox/exec/Unnest.cpp:186]: Nice refactor.

File: velox/exec/Unnest.cpp [hunk 11]
```diff
@@ -220,13 +220,35 @@ RowVectorPtr Unnest::generateOutput(
...
       // created dictionary.
       auto unnestBaseArray = currentDecoded.base()->as<ArrayVector>();
-      outputs[outputsIndex++] = identityMapping
-          ? unnestBaseArray->elements()
-          : wrapChild(
-                numElements,
-                elementIndices,
-                unnestBaseArray->elements(),
-                nulls);
+      auto elements = unnestBaseArray->elements();
```
Comment (mbasmanova) [inline velox/exec/Unnest.cpp:223]: I noticed that this method is very long. Perhaps, refactor to extract some helper methods.

File: velox/exec/tests/utils/PlanBuilder.h [hunk 12]
```diff
@@ -890,10 +892,16 @@ class PlanBuilder {
...
+  /// columns, one for each child type of the row. However, with legacyUnnest,
+  /// array of rows (like other array types) are expanded to a single
+  /// column of the row type. This single column output name is suffixed with
+  /// '_e'.
   PlanBuilder& unnest(
       const std::vector<std::string>& replicateColumns,
       const std::vector<std::string>& unnestColumns,
-      const std::optional<std::string>& ordinalColumn = std::nullopt);
+      const std::optional<std::string>& ordinalColumn = std::nullopt,
+      const std::optional<bool>& legacyUnnest = std::nullopt);
```
Comment (mbasmanova) [inline velox/exec/tests/utils/PlanBuilder.h:904]: Why do we need std::optional here? Can we just use bool?

File: velox/exec/tests/utils/PlanBuilder.cpp [hunk 13]
```diff
@@ -1430,11 +1432,22 @@ PlanBuilder& PlanBuilder::unnest(
...
     auto input = planNode_->outputType()->findChild(name);
     if (input->isArray()) {
-      unnestNames.push_back(name + "_e");
+      if (input->childAt(0)->isRow() && !isLegacyUnnest) {
+        // The array of rows is unnested into multiple columns, one for each
+        // child type of the row.
+        auto row = asRowType(input->childAt(0));
+        auto size = row->size();
+        for (auto i = 0; i < size; i++) {
+          unnestNames.push_back(row->nameOf(i) + "_e");
```
Comment (mbasmanova) [inline velox/exec/tests/utils/PlanBuilder.cpp:1446]: Should we include 'name' as well? What if there are multiple columns of the same array(row) type?

File: velox/exec/tests/UnnestTest.cpp [hunk 14]
```diff
@@ -39,6 +39,39 @@ TEST_F(UnnestTest, basicArray) {
 
   auto op = PlanBuilder().values({vector}).unnest({"c0"}, {"c1"}).planNode();
   assertQuery(op, "SELECT c0, UNNEST(c1) FROM tmp WHERE c0 % 7 > 0");
+
+  // Test with array containing dictionary encoded elements.
```
Comment (mbasmanova) [inline velox/exec/tests/UnnestTest.cpp:43]: This test is pretty hard to read and maintain. If we want to test encodings, it is better add a test that uses Fuzzer to generate inputs.

File: velox/exec/tests/UnnestTest.cpp [hunk 15]
```diff
@@ -103,6 +136,160 @@ TEST_F(UnnestTest, arrayWithOrdinality) {
...
+       }),
+       makeNullableFlatVector<int64_t>({-1, -2, -3, std::nullopt, -1, -2, -3}),
+       makeNullableFlatVector<double>(
+           {1.0, 2.0, 3.0, std::nullopt, 1.0, 2.0, 3.0}),
+       makeNullableFlatVector<StringView>(
+           {"cat", "tiger", "lion", std::nullopt, "cat", "tiger", "lion"})});
+  assertQuery(op, expected);
+}
+
+TEST_F(UnnestTest, encodedArrayOfRows) {
```
Comment (mbasmanova) [inline velox/exec/tests/UnnestTest.cpp:228]: Let's write a fuzzer-based test. Use Fuzzer to generate inputs, then run Unnest plan on these and on flattened copies. Compare the results.

**Review by mbasmanova — COMMENTED** [review]
@aditi-pandit Aditi, thank you for working on this. Overall looks good % a comment about parameter name, another comment about output of UnnestNode::toString() and some comments about tests.

File: velox/docs/develop/operators.rst (hunk 9 above)
Comment (aditi-pandit) [reply velox/docs/develop/operators.rst:674]: @mbasmanova : legacyUnnest unnests array of rows as a single column of row type. (!legacyUnnest) unnests as multiple columns of the child row types. So I presume you are suggesting flattenArrayOfRow or unnestArrayOfRow for the latter behavior (!legacyUnnest) (Your comment says replace legacyUnnest with those)

File: velox/core/PlanNode.h (hunk 2 above)
Comment (aditi-pandit) [reply velox/core/PlanNode.h:1933]: Using unnestArrayOfRows now to control the output instead.

File: velox/exec/Unnest.cpp (hunk 11 above)
Comment (aditi-pandit) [reply velox/exec/Unnest.cpp:223]: Created https://github.com/facebookincubator/velox/pull/8077 for the overall refactoring. I will move this logic into it.

**Comment by aditi-pandit** [comment]
@mbasmanova : Have addressed all the comments. PTAL. https://github.com/facebookincubator/velox/pull/8077 is separated for refactoring the long method in Unnest logic pointed in review.

**Comment by mbasmanova** [comment]
@aditi-pandit CI is red. Would you try rebasing?

File: velox/core/PlanNode.h [hunk 16]
```diff
@@ -1897,29 +1897,60 @@ class LimitNode : public PlanNode {
...
   /// @param unnestNames Names to use for unnested outputs: one name for each
-  /// array (element); two names for each map (key and value). The output names
-  /// must appear in the same order as unnestVariables.
+  /// array (element) except for array (row), which would have as many names
+  /// as the cardinality of the row type; two names for each map (key and
+  /// value).
+  /// The output names must appear in the same order as unnestVariables.
   /// @param ordinalityName Optional name for the ordinality columns. If not
   /// present, ordinality column is not produced.
+  /// @param unnestArrayOfRows Expands array of rows to multiple columns, one
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1923]: nit: If true, expands .... Otherwise, ...

File: velox/core/PlanNode.h [hunk 17]
```diff
@@ -1897,29 +1897,60 @@ class LimitNode : public PlanNode {
...
+      const std::vector<std::string>& unnestNames,
+      const std::optional<std::string>& ordinalityName,
+      const PlanNodePtr& source)
+      : UnnestNode(
+            id,
+            replicateVariables,
+            unnestVariables,
+            unnestNames,
+            ordinalityName,
+            true,
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1950]: Curious if default should be false. Doesn't current logic match unnestArrayOfRows = false?

File: velox/docs/develop/operators.rst [hunk 18]
```diff
@@ -665,9 +667,11 @@ number starting with 1.
    * - unnestVariables
      - Input columns of type array or map to expand.
    * - unnestNames
-     - Names to use for expanded columns. One name per array column. Two names per map column.
+     - Names to use for expanded columns. One name per array column. Two names per map column. If array of rows (and not legacyUnnest), then one column per child type of the row.
```
Comment (mbasmanova) [inline velox/docs/develop/operators.rst:670]: > legacyUnnest

Should we update this?

File: velox/exec/tests/PlanNodeToStringTest.cpp [hunk 19]
```diff
@@ -560,6 +560,17 @@ TEST_F(PlanNodeToStringTest, unnest) {
...
+
+  plan = PlanBuilder()
+             .values({data_})
+             .project({"array_constructor(c0) AS a0", "c1"})
+             .unnest({"c1"}, {"a0"}, std::nullopt, true)
+             .planNode();
+
+  ASSERT_EQ("-- Unnest\n", plan->toString());
+  ASSERT_EQ(
+      "-- Unnest[a0, unnest arrays of rows] -> c1:INTEGER, a0_e:SMALLINT\n",
```
Comment (mbasmanova) [inline velox/exec/tests/PlanNodeToStringTest.cpp:572]: Should this example use array(row) and not array(smallint)?

File: velox/exec/tests/QueryAssertionsTest.cpp [hunk 20]
```diff
@@ -493,4 +494,23 @@ TEST_F(QueryAssertionsTest, intervalDayTime) {
   assertQuery(plan, "SELECT * FROM tmp");
 }
 
+TEST_F(QueryAssertionsTest, equalPlans) {
```
Comment (mbasmanova) [inline velox/exec/tests/QueryAssertionsTest.cpp:497]: What is this test for?

File: velox/exec/tests/UnnestTest.cpp [hunk 21]
```diff
@@ -39,6 +40,19 @@ TEST_F(UnnestTest, basicArray) {
 
   auto op = PlanBuilder().values({vector}).unnest({"c0"}, {"c1"}).planNode();
   assertQuery(op, "SELECT c0, UNNEST(c1) FROM tmp WHERE c0 % 7 > 0");
+
+  // Tests with an array with dictionary encoded elements.
+  VectorFuzzer::Options opts;
+  opts.containerVariableLength = false;
```
Comment (mbasmanova) [inline velox/exec/tests/UnnestTest.cpp:46]: Why false? Usually, these kinds of tests have a loop with 100 or so iterations and allow Fuzzer to generate whatever encoding it chooses using fuzzInputRow.

File: velox/exec/tests/UnnestTest.cpp [hunk 22]
```diff
@@ -103,6 +117,146 @@ TEST_F(UnnestTest, arrayWithOrdinality) {
...
+            "lion",
+            "elephant",
+            "wolf",
+            std::nullopt,
+            std::nullopt})});
+  assertQuery(op, expected);
+
+  // Test with array of rows (one of which is a constant). Each array is
+  // encoded.
+  vector_size_t size = array1.size();
```
Comment (mbasmanova) [inline velox/exec/tests/UnnestTest.cpp:179]: Let's use fuzzer.

File: velox/exec/tests/UnnestTest.cpp [hunk 23]
```diff
@@ -103,6 +117,146 @@ TEST_F(UnnestTest, arrayWithOrdinality) {
...
+
+  auto unnestPlan = [&](const VectorPtr& input) -> const core::PlanNodePtr {
+    return PlanBuilder()
+        .values({makeRowVector({fuzzer.fuzzArray(input, 20)})})
+        .unnest({}, {"c0"}, std::nullopt, true)
+        .orderBy({"c0_c0_e", "c0_c1_e", "c0_c2_e"}, false)
+        .planNode();
+  };
+
+  auto noUnnestPlan = [&](const VectorPtr& child1,
```
Comment (mbasmanova) [inline velox/exec/tests/UnnestTest.cpp:223]: We should use the same plans but first use data generated by Fuzzer, then use same data but flattened. The results should be the same.

File: velox/exec/tests/utils/PlanBuilder.cpp [hunk 24]
```diff
@@ -1426,8 +1426,10 @@ PlanBuilder& PlanBuilder::nestedLoopJoin(
 PlanBuilder& PlanBuilder::unnest(
     const std::vector<std::string>& replicateColumns,
     const std::vector<std::string>& unnestColumns,
-    const std::optional<std::string>& ordinalColumn) {
+    const std::optional<std::string>& ordinalColumn,
+    const bool unnestArrayOfRows) {
```
Comment (mbasmanova) [inline velox/exec/tests/utils/PlanBuilder.cpp:1430]: remove const

File: velox/exec/tests/utils/QueryAssertions.h [hunk 25]
```diff
@@ -201,6 +201,11 @@ bool assertEqualResults(
     const std::vector<RowVectorPtr>& expected,
     const std::vector<RowVectorPtr>& actual);
 
+/// Ensure both plans have the same results.
+bool assertEqualResults(
```
Comment (mbasmanova) [inline velox/exec/tests/utils/QueryAssertions.h:205]: Maybe extract this change into a separate PR.

Comment (aditi-pandit) [reply velox/exec/tests/utils/QueryAssertions.h:205]: Sure. https://github.com/facebookincubator/velox/pull/8134
